### PR TITLE
Armv7a Arch code refactoring 

### PR DESCRIPTION
**Pull request by kishore-sn**
When we ported armv7a code from nuttx, we have not picked some of the arch related changes done in tizenrt for armv8m and other architectures. his commit picks these changes to arnv7a.

    - Port tizenrt specific changes from v8m to v7a
    - Move duplicate files in v7a to common directory
    - Rename some arch files (example arm_internal.h to up_internal.h)
    - Refactor and enhance assert code to handle app reload, dumps, etc..
    - Add configs to simplify the dumps which will be printed during assert
    - Enable app reloading and reload test scenarios
    - Fix heap corruption during app reload

### CODE REVIEW

File: build/configs/rtl8730e/loadable_apps_16M/defconfig
Comment (sunghan-chang) [inline build/configs/rtl8730e/loadable_apps_16M/defconfig:1278]: There is no mpu on rtl8730e. This looks strange. How about renaming as memory protection test regardless of this PR? I mean new PR.

File: os/arch/Kconfig
Comment (amogh-samsung) [inline os/arch/Kconfig:454]: Minor: commit title says - 
"This commit picks these changes to arnv7a" -> "armv7a" please

File: build/configs/rtl8730e/loadable_apps_16M/defconfig
Comment (kishore-sn) [reply build/configs/rtl8730e/loadable_apps_16M/defconfig]: Done. I added a commit for this.

File: os/board/rtl8730e/src/component/soc/amebad2/fwlib/ap_core/ameba_app_start.c
Comment (abhishek-samsung) [inline os/board/rtl8730e/src/component/soc/amebad2/fwlib/ap_core/ameba_app_start.c:40]: How about using the newly added config CONFIG_NUM_L2_PER_APP here so that user need not change this whenever the config is changed?

File: os/binfmt/binfmt_unloadmodule.c
Comment (abhishek-samsung) [inline os/binfmt/binfmt_unloadmodule.c:200]: Why is this assignment required? Isn't kmm_free enough?

Comment (kishore-sn) [reply os/binfmt/binfmt_unloadmodule.c:200]: We seem to be freeing these section multiple times based on whether the section pointer is non zero. Even though this wont create any issue, its better to always set the pointer to null after freeing it to avoid any future possibility of dangling pointers.

File: os/board/rtl8730e/src/component/soc/amebad2/fwlib/ap_core/ameba_app_start.c
Comment (kishore-sn) [reply os/board/rtl8730e/src/component/soc/amebad2/fwlib/ap_core/ameba_app_start.c:40]: I tried this, but this is causing several other issues. 
Even if we make this change, there are many other things like page size, etc which are defined in other header files. 
The main problem is that these values are defined in mmu.h and used in other assembly files of the target chipset. So, any attempt to use complex definition directives is resulting in multiple errors. So let us keep the current status.

File: os/kernel/task/task_dumptasks.c
Comment (sunghan-chang) [inline os/kernel/task/task_dumptasks.c:142]: With this change, TRAP is working? Could you check this with @thapav ?

File: os/binfmt/binfmt_arch_apis.c
Comment (sunghan-chang) [inline os/binfmt/binfmt_arch_apis.c:114]: os/kernel/task/task_terminate_unloaded.c
os/kernel/task/task_terminate.c

Both uses `CONFIG_ARM_MPU` instead of 7m_mpu and 8m_mpu configs.
https://github.com/Samsung/TizenRT/pull/5948/files#diff-de18ff5621c82ce73f2ccd2b0d9bbead6e92ae7836171469f03d4610b0cd6d34R191
Do we need to use them differently?

File: os/arch/arm/src/common/up_usestack.c
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_usestack.c:164]: Please check alignment

File: os/arch/arm/src/common/up_usestack.c
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_usestack.c:140]: This is a common file. It looks not good to add archi specific code here.
How about defining new config for this?
1. new config to on / off for -4
```
in up_userstack.c
#ifdef CONFIG_NEW
   top_.._stack = ....;
#else
   top_.._stack = .... - 4;
#endif

in Kconfig
config ARMV7A
    select NEW
```
2. new config to give adj value
```
in c
    top_...stack = .... + stack_size - CONFIG_NEW;

in defconfig
CONFIG_NEW=4
```

Comment (sunghan-chang) [reply os/arch/arm/src/common/up_usestack.c:140]: And I can't find why -4 is unnecessary. Could you leave comment?

**Review by sunghan-chang — COMMENTED** [review]
1. I can find mixed indentations, tab or spaces. We should use tabs instead of spaces. After this PR, please make a commit for them.
2. There are several modifications in common files. Are they verified on other chipsets?

**Comment by sunghan-chang** [comment]
@kishore-sn As you know, I am a beginner in this topic. Could you get review from your team and let me know result?

File: os/kernel/task/task_dumptasks.c
Comment (thapav) [reply os/kernel/task/task_dumptasks.c:142]: Sure I will check it once merged.

File: os/arch/arm/src/common/up_usestack.c
Comment (kishore-sn) [reply os/arch/arm/src/common/up_usestack.c:140]: hi @sunghan-chang 
I have been looking for a reason for this. But, as of now I have not been able to find out the exact reason. 
However, it looks like all cortex m code requires the -4 and cortex a32 does not need the -4.
If I try to change this in anyway, it is resulting in various crashes. 
So, I would like to keep this code for now and we will look into the reasoning for this at a later time. I think this can be a low priority activity.

Comment (kishore-sn) [reply os/arch/arm/src/common/up_usestack.c:140]: > This is a common file. It looks not good to add archi specific code here. How about defining new config for this?
> 
> 1. new config to on / off for -4
> 
> ```
> in up_userstack.c
> #ifdef CONFIG_NEW
>    top_.._stack = ....;
> #else
>    top_.._stack = .... - 4;
> #endif
> 
> in Kconfig
> config ARMV7A
>     select NEW
> ```
> 
> 2. new config to give adj value
> 
> ```
> in c
>     top_...stack = .... + stack_size - CONFIG_NEW;
> 
> in defconfig
> CONFIG_NEW=4
> ```

I am not sure what to name the CONFIG_NEW because I am not able to find out the reason for this.

**Comment by kishore-sn** [comment]
> 2. There are several modifications in common files. Are they verified on other chipsets?

@sunghan-chang 
I have verified build_test with these changes. 
Also, I have tested basic functionality on all rtk chipsets.

```
RTK8721
Flat bootup			: PASS
Loadable bootup		: PASS
app reload			: PASS
kernel tc 			: PASS

RTK 8720  (Same behavior without my commits also)
Flat bootup			: PASS
Loadable bootup		: FAIL (Hang)
app reload			: FAIL (Not enabled)
kernel tc 			: FAIL (Assertion)

RTK 8730
Flat bootup			: PASS
Loadable bootup		: PASS
app reload			: PASS
kernel tc 			: PASS
```

Please let me know if anything else needs to be tested

File: os/arch/arm/src/common/up_internal.h
Comment (ewoodev) [inline os/arch/arm/src/common/up_internal.h:426]: It make increase code in up_internal.h

up_XXX and arm_XXX both exist.
I think, we need to discuss about this.

1. keep arm_xxx and up_xxx, and remove common/internal.h and seperate per armvX-X
2. unify both into one, only in arch/arm.

File: os/arch/arm/src/common/up_internal.h
Comment (ewoodev) [inline os/arch/arm/src/common/up_internal.h:140]: This CONFIG_STACK_ALIGNMENT and STACK_ALIGNMENT also exist.

Comment (kishore-sn) [reply os/arch/arm/src/common/up_internal.h:140]: Removed it as it is not required.

File: os/arch/arm/src/common/up_internal.h
Comment (kishore-sn) [reply os/arch/arm/src/common/up_internal.h:426]: True. We need to discuss this. 
But let us merge this PR first, as it is required to verify many features on 8730.
We can take up this activity separately.

**Review by sunghan-chang — DISMISSED** [review]

File: os/arch/Kconfig
Comment (sunghan-chang) [inline os/arch/Kconfig:457]: Please check first commit, it's another commit and it has your changes. Something wrong.

File: os/arch/arm/src/armv7-a/arm_mmu.c
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-a/arm_mmu.c:517]: Will you handle this for the TRAP later? @thapav

Comment (kishore-sn) [reply os/arch/arm/src/armv7-a/arm_mmu.c:517]: Yes Chang.. We will make sure to raise a separate PR for TRAP.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**
